backend/src/main/java/de/sambam/gamenighttracker/service/UserService.java

```diff
@@ -23,73 +23,84 @@ public UserService(UserDb userDb) {
     }
 
-    public List<PlayerDto> listAllPlayers(String userId) {
-        Optional<User> user = userDb.findById(userId);
+
+    public List<PlayerDto> listAllPlayers(String username) {
+        List<Game> gameList = getGamesForUser(username);
         Map<String, PlayerDto> playerList = new HashMap<>();
+        for (Game game : gameList) {
+            getPlayersForOneGame(game).forEach(player ->
+                    playerList.put(player.getName(), PlayerDto.builder()
+                            .name(player.getName())
+                            .color(player.getColor())
+                            .build()));
+        }
+        return playerList.values().stream().collect(Collectors.toList());
+    }
 
-        if (user.isPresent()) {
-            List<Game> gameList = user.get().getPlayedGames();
-            if (gameList != null) {
-                for (Game game : gameList) {
-                    if (game != null) {
-                        List<GameSession> existingSessionList = game.getGameSessionList();
-                        if (existingSessionList != null) {
-                            for (GameSession session : existingSessionList) {
-                                if (session != null) {
-                                    List<Player> existingPlayerList = session.getPlayerList();
-                                    if (existingPlayerList != null)
-                                        for (Player player : existingPlayerList) {
-                                            if (player != null) {
-                                                playerList.put(player.getName(), PlayerDto.builder()
-                                                        .name(player.getName())
-                                                        .color(player.getColor())
-                                                        .build());
-
-                                            }
-                                        }
-                                }
-                            }
+    private List<Player> getPlayersForOneGame(Game game) {
+        List<Player> playerList = new ArrayList<>();
+
+        if (game == null) {
+            return playerList;
+        }
+        List<GameSession> existingSessionList = game.getGameSessionList();
+        if (existingSessionList == null) {
+            return playerList;
+        }
+        for (GameSession session : existingSessionList) {
+            if (session != null) {
+                List<Player> existingPlayerList = session.getPlayerList();
+                if (existingPlayerList != null)
+                    for (Player player : existingPlayerList) {
+                        if (player != null) {
+                            playerList.add(player);
                         }
                     }
-                }
             }
         }
-        return playerList.values().stream().collect(Collectors.toList());
+        return playerList;
```

Comment:
Separate the control flow to own methods and use `stream` instead. This should significantly increase comprehensibility. I hope so :)